### PR TITLE
[Feature Addition] Mish Activation function

### LINKED ISSUE #4797: Feature Request: Add Mish Activation
Mish is a new novel activation function proposed in this [paper](https://arxiv.org/abs/1908.08681).
It has shown promising results so far and has been adopted in several packages including:

- [TensorFlow-Addons](https://github.com/tensorflow/addons/tree/master/tensorflow_addons/activations)
- [SpaCy (Tok2Vec Layer)](https://github.com/explosion/spaCy)
- [Thinc - SpaCy's official NLP based ML library](https://github.com/explosion/thinc/releases/tag/v7.3.0)
- [Echo AI](https://github.com/digantamisra98/Echo)
- [Eclipse's deeplearning4j](https://github.com/eclipse/deeplearning4j/issues/8417)
- [CNTKX - Extension of Microsoft's CNTK](https://github.com/delzac/cntkx)
- [FastAI-Dev](https://github.com/fastai/fastai_dev/blob/0f613ba3205990c83de9dba0c8798a9eec5452ce/dev/local/layers.py#L441)
- [Darknet](https://github.com/AlexeyAB/darknet/commit/bf8ea4183dc265ac17f7c9d939dc815269f0a213)
- [Yolov3](https://github.com/ultralytics/yolov3/commit/444a9f7099d4ff1aef12783704e3df9a8c3aa4b3)
- [BeeDNN - Library in C++](https://github.com/edeforas/BeeDNN)
- [Gen-EfficientNet-PyTorch](https://github.com/rwightman/gen-efficientnet-pytorch)
- [dnet](https://github.com/umangjpatel/dnet/blob/master/nn/activations.py)

All benchmarks, analysis and links to official package implementations can be found in this [repository](https://github.com/digantamisra98/Mish) 

It would be nice to have Mish as an option within the activation function group.

This is the comparison of Mish with other conventional activation functions in a SEResNet-50 for CIFAR-10:  (Better accuracy and faster than GELU)
se50_1

### PR DESCRIPTION
**Pull request by thebhatman**
This PR resolves https://github.com/shogun-toolbox/shogun/issues/4797. Currently I have added the forward pass computation of the Mish Function.

### CODE REVIEW

File: src/shogun/neuralnets/NeuralMishLayer.cpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
+/*
```
Comment (gf712) [inline src/shogun/neuralnets/NeuralMishLayer.cpp:1]: You can switch the copyright notice to:
```
/*
 * This software is distributed under BSD 3-clause license (see LICENSE file).
 *
 * Authors: ...
 */
```
Keeps the files less verbose

File: tests/unit/neuralnets/NeuralMishLayer_unittest.cc [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+ * activations
+ */
+TEST_F(NeuralMishLayerTest, compute_activations)
+{
+	int32_t seed = 100;
+	// initialize some random inputs
+	SGMatrix<float64_t> x;
+	std::shared_ptr<NeuralInputLayer> input;
+	std::mt19937_64 prng(seed);
+	std::tie(x, input) = setup_input_layer<float64_t>(12, 3, -10.0, 10.0, prng);
```
Comment (gf712) [inline tests/unit/neuralnets/NeuralMishLayer_unittest.cc]: you could use structured bindings here, `auto [x, input] = ...`, as we are now on C++17

File: tests/unit/neuralnets/NeuralMishLayer_unittest.cc [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	{
+		A_ref[idx] = A_ref[idx] * std::tanh(std::log(1 + std::exp(A_ref[idx])));
+	}
+
+	// compare
+	EXPECT_EQ(A_ref.num_rows, A.num_rows);
+	EXPECT_EQ(A_ref.num_cols, A.num_cols);
+	for (int32_t i = 0; i < A.num_rows * A.num_cols; i++)
+	{
+		EXPECT_NEAR(A_ref[i], A[i], 1e-12);
```
Comment (gf712) [inline tests/unit/neuralnets/NeuralMishLayer_unittest.cc]: the 1e-12 is a bit arbitrary, maybe use std::numeric_limits<float64_t>::epsilon?

File: src/shogun/neuralnets/NeuralMishLayer.cpp [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+{
+}
+
+void NeuralMishLayer::compute_activations(SGVector<float64_t> parameters,
+    const std::vector<std::shared_ptr<NeuralLayer>>& layers)
+{
+	NeuralLinearLayer::compute_activations(parameters, layers);
+
+	int32_t len = m_num_neurons*m_batch_size;
+	for (int32_t i=0; i<len; i++)
```
Comment (gf712) [inline src/shogun/neuralnets/NeuralMishLayer.cpp]: you could use a ranged loop here no?
```cpp
for (auto& activation: m_activations)
{
  activation *= std::tanh(std::log(1 + std::exp(activation)));
}

Comment (gf712) [reply src/shogun/neuralnets/NeuralMishLayer.cpp]: or even better, std::transform with the same input and output iterators, that way it is clear that you are changing some values

File: tests/unit/neuralnets/NeuralMishLayer_unittest.cc (hunk 2 above)
Comment (gf712) [reply tests/unit/neuralnets/NeuralMishLayer_unittest.cc]: also maybe create some variables for those parameters, and use those instread of x.num_cols, layer->get_num_neurons(), and so on.. It just makes it easier to read the code imo

File: tests/unit/neuralnets/NeuralMishLayer_unittest.cc [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+/** Compares the activations computed using the layer against manually computed
+ * activations
+ */
+TEST_F(NeuralMishLayerTest, compute_activations)
+{
+	int32_t seed = 100;
+	// initialize some random inputs
+	SGMatrix<float64_t> x;
+	std::shared_ptr<NeuralInputLayer> input;
```
Comment (gf712) [inline tests/unit/neuralnets/NeuralMishLayer_unittest.cc]: these lines can go no?
```
SGMatrix<float64_t> x;
std::shared_ptr<NeuralInputLayer> input;
```

File: src/shogun/neuralnets/NeuralMishLayer.cpp [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+NeuralLinearLayer(num_neurons)
+{
+}
+
+void NeuralMishLayer::compute_activations(SGVector<float64_t> parameters,
+    const std::vector<std::shared_ptr<NeuralLayer>>& layers)
+{
+	NeuralLinearLayer::compute_activations(parameters, layers);
+
+	int32_t len = m_num_neurons*m_batch_size;
```
Comment (gf712) [inline src/shogun/neuralnets/NeuralMishLayer.cpp]: this is not needed anymore

File: tests/unit/neuralnets/NeuralMishLayer_unittest.cc (hunk 5 above)
Comment (thebhatman) [reply tests/unit/neuralnets/NeuralMishLayer_unittest.cc]: Yeah, thanks for pointing out. Will change it.

Comment (gf712) [reply tests/unit/neuralnets/NeuralMishLayer_unittest.cc]: no worries, sorry for being so pedantic :D

**Comment by thebhatman** [comment]
Also should I implement a custom gradient for Mish? Can you give me some pointers on how I should be implementing the gradient?

**Comment by gf712** [comment]
> Also should I implement a custom gradient for Mish? Can you give me some pointers on how I should be implementing the gradient?

I am not sure how to do it, I have never written neural net stuff for shogun. But we don't have any support for autodiff right now so I am assuming you write it manually, like for softmax:
https://github.com/shogun-toolbox/shogun/blob/7db6e4ef151ed94154ec5db326eedf7fb7368ba4/src/shogun/neuralnets/NeuralSoftmaxLayer.h#L74-L84

maybe @karlnapf knows what needs to be done?

**Comment by thebhatman** [comment]
Thank you! I have added local gradients for Mish. Currently I have defined gradients for when Mish is used in the hidden layers. At the output layer, I haven't changed the gradient as in both ReLU and Softmax, the gradient at output layer was same and Mean squared loss was assumed.. Please let me know if I should change the gradient at output layer.

File: src/shogun/neuralnets/NeuralMishLayer.cpp [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+void NeuralMishLayer::compute_local_gradients(
+		SGMatrix<float64_t> targets)
+{
+	if (targets.num_rows != 0)
+	{
+		int32_t length = m_num_neurons*m_batch_size;
+		for (int32_t i=0; i<length; i++)
+		{
+			m_local_gradients[i] = (m_activations[i]-targets[i])/m_batch_size;
```
Comment (karlnapf) [inline src/shogun/neuralnets/NeuralMishLayer.cpp:40]: this should be vectorized

Comment (karlnapf) [reply src/shogun/neuralnets/NeuralMishLayer.cpp:40]: also this code seems to be re-usable across different activation functions ... why have it down in here?

File: src/shogun/neuralnets/NeuralMishLayer.cpp [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+		}
+	}
+	else
+	{
+		int32_t len = m_num_neurons*m_batch_size;
+		for (int32_t i=0; i< len; i++)
+		{
+			auto e = std::exp(m_activations[i]);
```
Comment (karlnapf) [inline src/shogun/neuralnets/NeuralMishLayer.cpp:48]: make sure to put a unit test that checks for correctness of this manually implemented gradient on a few simple non trivial cases...

Comment (karlnapf) [reply src/shogun/neuralnets/NeuralMishLayer.cpp:48]: again, could this be vectorized?

**Review by karlnapf — COMMENTED** [review]
Thanks for the patch! I think it could be made a bit faster via avoiding explicit.

File: src/shogun/neuralnets/NeuralMishLayer.cpp (hunk 7 above)
Comment (thebhatman) [reply src/shogun/neuralnets/NeuralMishLayer.cpp:40]: I had a doubt whether I had to implement gradient at the output layer separately. The other activation functions also had this. So I put it there. I'll remove it if it's not required.

File: src/shogun/neuralnets/NeuralMishLayer.cpp (hunk 8 above)
Comment (thebhatman) [reply src/shogun/neuralnets/NeuralMishLayer.cpp:48]: I'm not really sure how to vectorize this. It would be really helpful if you could point me to a similar example where vectorized operations have been used....

Comment (gf712) [reply src/shogun/neuralnets/NeuralMishLayer.cpp:48]: by vectorized he means you could/should use Eigen. You can map the memory of the matrix to an eigen matrix and use their implementation of coefficient wise transformation.
Look for example at 
https://github.com/shogun-toolbox/shogun/blob/895760f011c5833cd290192f37bdef1457e0ebf8/src/shogun/mathematics/linalg/backend/eigen/BasicOps.cpp#L233-L243
This will compile with SIMD instructions if they are available and inline properly

Comment (thebhatman) [reply src/shogun/neuralnets/NeuralMishLayer.cpp:48]: I am thinking of adding an implementation of Mish here; something similar to this : https://github.com/shogun-toolbox/shogun/blob/7db6e4ef151ed94154ec5db326eedf7fb7368ba4/src/shogun/mathematics/linalg/backend/eigen/SpecialPurposes.cpp#L162

Comment (thebhatman) [reply src/shogun/neuralnets/NeuralMishLayer.cpp:48]: Would this be correct?

Comment (gf712) [reply src/shogun/neuralnets/NeuralMishLayer.cpp:48]: yes, you could do that!

**Comment by karlnapf** [comment]
What's the state of this thing? Sorry for being a bit slow on our side atm

File: src/shogun/mathematics/linalg/backend/eigen/SpecialPurposes.cpp [hunk 9]
```diff
@@ -157,6 +157,18 @@ void LinalgBackendEigen::rectified_linear_impl(
...
 
+template <typename T>
+void LinalgBackendEigen::mish_impl(
+    const SGMatrix<T>& a, SGMatrix<T>& result) const
+{
+	typename SGMatrix<T>::EigenMatrixXtMap a_eig = a;
+	typename SGMatrix<T>::EigenMatrixXtMap result_eig = result;
+
+	for (index_t i = 0; i < a_eig.rows() * a_eig.cols(); ++i)
+		result_eig(i) = a_eig(i) * std::tanh(std::log(1 + std::exp(a_eig(i))));
```
Comment (gf712) [inline src/shogun/mathematics/linalg/backend/eigen/SpecialPurposes.cpp:168]: can you not write this as something like `result_eig = a_eig.array() * ((1 + a_eig.exp()).log()).tanh()`? That way you give eigen the best chance to vectorise this.

**Comment by karlnapf** [comment]
Hi! Shall we try to finish this up? :)

**Comment by digantamisra98** [comment]
Yeah @karlnapf. Sorry for the inactivity. Let me catch up with the commits again.

File: src/shogun/neuralnets/NeuralMishLayer.cpp (hunk 4 above)
Comment (YashasSamaga) [reply src/shogun/neuralnets/NeuralMishLayer.cpp]: `std::log1p` instead of `std::log(1 + std::exp)` is better numerical accuracy but is still poor.

OpenCV CPU backend has nearly 2x faster and even more accurate [implementation](https://github.com/opencv/opencv/blob/4564b8a224545a54c545e9d2743cce3c0e1f1c5b/modules/dnn/src/layers/elementwise_layers.cpp#L760-L773).

File: src/shogun/mathematics/linalg/backend/eigen/SpecialPurposes.cpp (hunk 9 above)
Comment (YashasSamaga) [reply src/shogun/mathematics/linalg/backend/eigen/SpecialPurposes.cpp:168]: `log1p` instead of `log(1 + exp(x))` is better for numerical accuracy.

OpenCV CPU backend has a much faster (more than 2x) and accurate [implementation](https://github.com/opencv/opencv/blob/4564b8a224545a54c545e9d2743cce3c0e1f1c5b/modules/dnn/src/layers/elementwise_layers.cpp#L760-L773).

File: src/shogun/neuralnets/NeuralMishLayer.cpp (hunk 4 above)
Comment (gf712) [reply src/shogun/neuralnets/NeuralMishLayer.cpp]: do you know what the accuracy tradeoff is?

File: src/shogun/mathematics/linalg/backend/eigen/SpecialPurposes.cpp (hunk 9 above)
Comment (gf712) [reply src/shogun/mathematics/linalg/backend/eigen/SpecialPurposes.cpp:168]: do you know what the accuracy tradeoff is?

Comment (YashasSamaga) [reply src/shogun/mathematics/linalg/backend/eigen/SpecialPurposes.cpp:168]: Errors callculated against values computed using `tanh(log1p(exp(x)))` with 128-bit floats.

errors

YOLOv4 performance almost doubles: https://github.com/opencv/opencv/pull/17624

Comment (karlnapf) [reply src/shogun/mathematics/linalg/backend/eigen/SpecialPurposes.cpp:168]: we could have the approximation somehwere, but it should be marked as such
(and I agree with @gf712 this can be vectorized, even the approximation)
